## 错误处理

假如现在有一个业务流程有 4 个步骤需要执行，每个步骤都有可能抛出错误

```go
func step1() error {}
func step2() error {}
func step3() error {}
func step4() error {}
```

### 1. 常用处理方式

```go
func main() {
	err := step1()
  if err != nil {
    fmt.Println("has error")
    return err
  }
  err := step2()
  if err != nil {
    fmt.Println("has error")
    return err
  }
  err := step3()
  if err != nil {
    fmt.Println("has error")
    return err
  }
  err := step4()
  if err != nil {
    fmt.Println("has error")
    return err
  }
}
```

在每一个步骤执行完成后判断 error。 这种方式最灵活，适用于业务逻辑不稳定的场景

### 2. 内置错误到步骤中

```go
type TestInterface interface {
	Step1()
	Step2()
	Step3()
	Step4()

  Error() err
}

type Test struct {
  err error
}

func (t *test)Step1() {
  if t.Error() != nil {
    return t.Error()
  }
}

func main() {
  t := &Test{}
  t.Step1()
  t.Step2()
  t.Step3()
  t.Step4()
  if t.Error() != nil {
    return t.Error()
  }
}
```

这种处理方式在主业务逻辑中只需要判断一次。 但是确定是逻辑比较固定，在每一个方法里还是要判断一次。

### 3. 利用函数式延迟执行

```go
func step1() error {}
func step2() error {}
func step3() error {}
func step4() error {}

type t interface {
  step() error
}

func BreakOnError(sl ) {

}
```

### 4. 利用接口特性全局保存 error

```go
type errorWriter struct {
	w   io.Writer
	err error
}

func (r *errorWriter) Write(buf []byte) (n int, err error) {
	if r.err != nil {
		return 0, err
	}

	n, err = r.w.Write(buf)
	if err != nil {
		r.err = err
		return
	}
	return
}

func test111(w io.Writer) error {
	ew := &errorWriter{
		w:   w,
		err: nil,
	}
	// 这里不需要判断err了
	_, _ = fmt.Fprintf(ew, "has error: %v", errors.New("11"))

	// 这里不需要判断err了
	_, _ = fmt.Fprintf(ew, "has error2: %v", errors.New("22"))

	return ew.err
}
```
